src/chunk.c

```diff
@@ -189,19 +196,96 @@ size_t Uncompressed_DelRange(Chunk_t *chunk, timestamp_t startTs, timestamp_t en
     return deleted_count;
 }
 
-void Uncompressed_ResetChunkIterator(ChunkIter_t *iterator, Chunk_t *chunk) {
+void Uncompressed_ResetChunkIterator(ChunkIter_t *iterator, const Chunk_t *chunk) {
     ChunkIterator *iter = (ChunkIterator *)iterator;
-    iter->chunk = chunk;
+    iter->chunk = (Chunk_t *)chunk;
     if (iter->options & CHUNK_ITER_OP_REVERSE) { // iterate from last to first
         iter->currentIndex = iter->chunk->num_samples - 1;
     } else { // iterate from first to last
         iter->currentIndex = 0;
     }
 }
 
-ChunkIter_t *Uncompressed_NewChunkIterator(Chunk_t *chunk,
+#define __array_reverse_inplace(arr, len)                                                          \
+    __extension__({                                                                                \
+        const size_t ei = len - 1;                                                                 \
+        __typeof__(*arr) tmp;                                                                      \
+        for (size_t i = 0; i < len / 2; ++i) {                                                     \
+            tmp = arr[i];                                                                          \
+            arr[i] = arr[ei - i];                                                                  \
+            arr[ei - i] = tmp;                                                                     \
+        }                                                                                          \
+    })
+
+void reverseDomainChunk(DomainChunk *domainChunk) {
+    __array_reverse_inplace(domainChunk->samples.timestamps, domainChunk->num_samples);
+    __array_reverse_inplace(domainChunk->samples.values, domainChunk->num_samples);
+    domainChunk->rev = true;
+}
+
+// TODO: can be optimized further using binary search
+DomainChunk *Uncompressed_ProcessChunk(const Chunk_t *chunk,
+                                       uint64_t start,
+                                       uint64_t end,
+                                       DomainChunk *domainChunk,
+                                       DomainChunk *domainChunkAux,
+                                       bool reverse) {
+    const Chunk *_chunk = chunk;
+    domainChunk->num_samples = 0;
+    if (unlikely(!_chunk || _chunk->num_samples == 0 || end < start ||
+                 _chunk->base_timestamp > end ||
+                 _chunk->samples[_chunk->num_samples - 1].timestamp < start)) {
+        return NULL;
+    }
+
+    size_t si = _chunk->num_samples, ei = _chunk->num_samples - 1, i = 0;
+
+    // find start index
+    for (; i < _chunk->num_samples; i++) {
+        if (_chunk->samples[i].timestamp >= start) {
+            si = i;
+            break;
+        }
+    }
+
+    if (si == _chunk->num_samples) { // all TS are smaller than start
+        return NULL;
+    }
+
+    // find end index
+    for (; i < _chunk->num_samples; i++) {
```

Comment:
you don't really need to find the end of the range, you can just merge this to the copy loop

Reply:
@danni-m I really feels like this is counter productive, since in the near future uncompressed chunk will become SOA (struct of arrays) and memcpy will have a better perfomance, so I think that, also probably compressed chunk logic will also use this function because we will use vectorized decompression.
So maybe it's ok to leave the current implementation to support such a scenario in the future.
I also don't think it's really hurt the simplicity of the code.